### PR TITLE
[REVIEW] Enable implicit casting when concatenating mixed types

### LINKED ISSUE #8264: [FEA] Implicit typecasting between Decimal and Floats during concat
**Is your feature request related to a problem? Please describe.**
Concat automatically handles typecasting between to decimal columns to decide the precision and scale of the output resulting decimal column during concat. There is implicit typecasting between other numeric column types as well (float32/64 and ints). However it isn't possible to concat between float and decimal columns in a dataframe.

**Describe the solution you'd like**
```python
df1['a'] = range(10)
df1['b'] = 1.234
df1.b = df1['b'].astype(Decimal64Dtype(7,2))

df2['b'] = range(10)
df2['a'] = 1.234
cudf.concat([df1, df2])
```
`ValueError: All columns must be the same type`

```python
df1['a'] = range(10)
df1['b'] = 1.234
df1.b = df1['b'].astype("float32")

df2['b'] = range(10)
df2['a'] = 1.234
cudf.concat([df1, df2])
# Works
```

**Describe alternatives you've considered**
Alternative is to cast explicitly before concat

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by ChrisJar**
This enables implicit casting when decimal columns are concatenated with numeric columns by casting the numeric columns to decimal columns.

Closes #8264

### CODE REVIEW

File: python/cudf/cudf/utils/dtypes.py [hunk 1]
```diff
@@ -291,14 +291,21 @@ def is_decimal_dtype(obj):
...
-    s = max([a.dtype.scale for a in args])
-    lhs = max([a.dtype.precision - a.dtype.scale for a in args])
-    p = min(cudf.Decimal64Dtype.MAX_PRECISION, s + lhs)
-    dtype = cudf.Decimal64Dtype(p, s)
-
+    dtype = _find_common_type_decimal([a.dtype for a in args])
     return [a.astype(dtype) for a in args]
 
 
+def _find_common_type_decimal(dtypes):
```
Comment (shwina) [inline python/cudf/cudf/utils/dtypes.py:298]: Can we call this in `find_common_type` [here](https://github.com/rapidsai/cudf/blob/6136fbc6d0657315a073cf175de0903d6d1bf272/python/cudf/cudf/utils/dtypes.py#L700)? And then avoid importing this underscore function in other modules?

File: python/cudf/cudf/core/series.py [hunk 2]
```diff
@@ -2402,10 +2404,30 @@ def _concat(cls, objs, axis=0, index=True):
...
+                    ]
+                ):
+                    decimal_type = _find_common_type_decimal(
+                        [
+                            obj.dtype
+                            for obj in objs
+                            if is_decimal_dtype(obj.dtype)
+                        ]
+                    )
+                    objs = [obj.astype(decimal_type) for obj in objs]
```
Comment (shwina) [inline python/cudf/cudf/core/series.py:2430]: Is it possible for all this logic to be handled in `find_common_type`? For example, it'd be nice to be able to replace this with just:

```python
common_dtype = find_common_type(objs)
objs = [obj.astype(common_type) for obj in objs
```

**Review by shwina — APPROVED** [review]
Looks great to me! Awesome refactoring of some of the dtype handling too!

**Review by brandon-b-miller — APPROVED** [review]

**Comment by shwina** [comment]
@gpucibot merge

**Pull request merged**